Joystream-Role-Bot/src/controls/control.ts

```diff
@@ -132,32 +154,52 @@ export const setMemberRole = async (client: Client): Promise<void> => {
     //   console.error("Error removing member roles:", error);
     // }
 
-    await discordMember.roles.add(RoleAddress.membershipLinked);
 
-    const roleUpdatePromises = qnMember.roles.map(async (memberRole) => {
-      const memberRoleGroupId = memberRole.groupId;
-      const mappedRoles = roleMap[memberRoleGroupId];
+    const discordMemberAllRolesIds = discordMember.roles.cache.map((role) => role.id)
+    const roleAddressArray: string[] = Object.values(RoleAddress);
+    const discordMemberRolesIds = discordMemberAllRolesIds.filter((item) => roleAddressArray.includes(item))
+
+    if (!discordMemberRolesIds.find((id: string) => id === RoleAddress.membershipLinked)) {
+      await discordMember.roles.add(linkRole);
+    }
+
+    const rolesMember = qnMember.roles.filter(group => group.status.__typename === "WorkerStatusActive");
+
+    let worker = false;
+
+    let activeRoles: string[] = [];
+
+    rolesMember.forEach(async (work) => {
+      const mappedRoles = roleMap[work.groupId];
+
       if (!mappedRoles) return;
 
       const [leadRoleId, workerRoleId] = mappedRoles;
-      const roleId = memberRole.isLead ? leadRoleId : workerRoleId;
+
+      const roleId = work.isLead ? leadRoleId : workerRoleId;
+
       const role = await guild.roles.fetch(roleId);
 
       if (!role) {
         console.log(`<@&${roleId}> Role not found`);
         return;
       }
+      if (!discordMemberRolesIds.find((id: string) => id === roleId)) discordMember.roles.add(role);
+      activeRoles.push(roleId);
+      worker = true;
+    })
+
+    await Promise.all(rolesMember);
+
+    const terminatedRoles = discordMemberRolesIds.filter(item => !activeRoles.includes(item))
+    terminatedRoles.forEach(id => discordMember.roles.remove(id));
+    await Promise.all(terminatedRoles);
```

Comment:
`terminatedRoles` are not promises, you should do `const terminatedRolesPromises = terminatedRoles.map` and await those